### PR TITLE
ARROW-15215: [C++] Consolidate kernel data-copy utilities between replace_with_mask, case_when, coalesce, choose, fill_null_forward, fill_null_backward

### PR DESCRIPTION
**Pull request by JabariBooker**
Consolidate kernel data-copy utilities between replace_with_mask, case_when, coalesce, choose, fill_null_forward, fill_null_backward in one place

### CODE REVIEW

**Comment by lidavidm** [comment]
Should this be marked as draft for now?

**Comment by JabariBooker** [comment]
> Should this be marked as draft for now?

Yes, for now

File: cpp/src/arrow/compute/kernels/copy_data.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data.h:1]: Can we name the header `copy_data_internal.h`? That way it won't get installed/exposed to outside users.

File: cpp/src/arrow/compute/kernels/copy_data.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/array/builder_nested.h"
+#include "arrow/array/builder_primitive.h"
+#include "arrow/array/builder_time.h"
+#include "arrow/array/builder_union.h"
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data.h:23]: Looks like all the builder headers are unused here.

File: cpp/src/arrow/compute/kernels/copy_data.h [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+#include "arrow/compute/api.h"
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data.h:24]: Do we need to include api.h? This includes a lot of headers.

File: cpp/src/arrow/compute/kernels/copy_data.h [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+#include "arrow/compute/kernels/codegen_internal.h"
+#include "arrow/util/bit_block_counter.h"
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data.h:26]: BitBlockCounter isn't used here.

File: cpp/src/arrow/compute/kernels/copy_data.h [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+#include "arrow/util/bit_run_reader.h"
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data.h:27]: Nor is BitRunReader.

File: cpp/src/arrow/compute/kernels/copy_data.h [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+#include "arrow/util/bitmap.h"
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data.h:28]: Don't think we use Bitmap either.

File: cpp/src/arrow/compute/kernels/copy_data.h [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/bitmap_reader.h"
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data.h:30]: Nor do we use any of the bitmap readers.

File: cpp/src/arrow/compute/kernels/copy_data.h [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  static void CopyData(const DataType&, uint8_t* out, const int64_t out_offset,
+                       const ArrayData& in, const int64_t in_offset,
+                       const int64_t length) {
+    const auto in_arr = in.GetValues<uint8_t>(1, /*absolute_offset=*/0);
+    arrow::internal::CopyBitmap(in_arr, in_offset + in.offset, length, out, out_offset);
+  }
+
+  static void CopyData(const DataType&, uint8_t* out, const int64_t out_offset,
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data.h:60]: Can we consolidate CopyData and CopyArray/CopyScalar? These four methods do basically the same things, just with slightly different signatures.

Comment (JabariBooker) [reply cpp/src/arrow/compute/kernels/copy_data.h]: I did some consolidation, but there are still 3 functions (all `CopyData`) in `CopyDataUtils`, one for each passed data representation: `Scalar`, `ArrayData`, and `const uint8*`

File: cpp/src/arrow/compute/kernels/copy_data_internal.h [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+namespace arrow {
+namespace compute {
+namespace internal {
+
+template <typename Type, typename Enable = void>
+struct CopyDataUtils {};
+
+template <>
+struct CopyDataUtils<BooleanType> {
+  static void CopyData(const DataType&, const Scalar& in, const int64_t in_offset,
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data_internal.h:31]: Do the Scalar overloads need in_offset?

File: cpp/src/arrow/compute/kernels/copy_data_internal.h [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+template <>
+struct CopyDataUtils<FixedSizeBinaryType> {
+  static void CopyData(const DataType& ty, const Scalar& in, const int64_t in_offset,
+                       uint8_t* out, const int64_t out_offset, const int64_t length) {
+    const int32_t width = checked_cast<const FixedSizeBinaryType&>(ty).byte_width();
+    uint8_t* begin = out + (width * out_offset);
+    const auto& scalar = checked_cast<const arrow::internal::PrimitiveScalarBase&>(in);
+    // Null scalar may have null value buffer
+    if (!scalar.is_valid) return;
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/copy_data_internal.h:58]: The original impl had a `std::memset` in this case to ensure the underlying array is still initialized.

**Review by lidavidm — COMMENTED** [review]
Thanks for the cleanup. Just a couple minor things.

File: cpp/src/arrow/compute/kernels/copy_data_internal.h (hunk 10 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/copy_data_internal.h:58]: Once we fix this we can trigger Conbench here and take a look at the benchmarks.

File: cpp/src/arrow/compute/kernels/copy_data_internal.h (hunk 9 above)
Comment (JabariBooker) [reply cpp/src/arrow/compute/kernels/copy_data_internal.h:31]: `ReplaceWithArrayMaskImpl` in `vector_replace.cc` makes calls to `CopyData` using `ArrayData` and `Scalar` for the `Data` template type. When trying to select for which `CopyData` overload to call, there are errors in compilation when trying to differentiate the `ArrayData` version of `CopyData` from the Scalar version.

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/copy_data_internal.h:31]: Ah, I see now. Thanks.

**Comment by lidavidm** [comment]
@ursabot please benchmark lang=C++

**Pull request closed without merging**